add support for AKS BYOCNI

nbusseneau opened the pull request:
Microsoft has recently released support for "Bring your own CNI" when creating AKS clusters, allowing users to create clusters with no CNI plugin pre-installed.

This is extremely useful for Cilium as previously setting up an AKS cluster and installing Cilium was very complicated, due to the need to use multiple nodepools with a complex taint system in order to ensure applications pods would not get scheduled before Cilium agents were ready to manage connectivity in the cluster.

The BYOCNI feature has been available in preview as an `az` CLI extension for some time now, and Microsoft has published a dedicated documentation page outlining the prequisites and implications (notably in terms of support policy) of BYOCNI: https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli

BYOCNI in its current state is a strong enough proposal to warrant a change to the recommendations we make to our users for newly created AKS clusters.

Here are the changes necessary to run Cilium on an AKS cluster in BYOCNI mode:

- Azure IPAM does not work in BYOCNI mode as the Azure API is not available. Instead of using the Azure operator, we must use the generic operator with Cluster Pool IPAM.
- Direct routing does not work in BYOCNI mode. Instead, we must use VXLAN encapsulation.

See per commit for a more thorough description of the implementation.

### ⚠️ Additional notes for reviewers

- See [this comment](https://github.com/cilium/cilium/pull/19379#issuecomment-1156611493) for example of working workflow run with temporary commit.

### ⚠️ Checklist

- [x] ~~We are currently not making any changes to the old Azure IPAM-based installation method, however it is currently unusable due to the taints not working on AKS (see https://github.com/Azure/AKS/issues/2934).~~ Instructions for Azure IPAM have been updated.
- [x] This PR depends on https://github.com/cilium/cilium-cli/pull/899 being merged and released, at which point we can add a commit in this PR to update the CLI version used in the AKS workflows.

### ⚠️ After merge

- [ ] Re-enable AKS CI workflows.

nbusseneau committed:
ci: switch to AKS BYOCNI in `master` CI

Following up on the previous commits: we switch to BYOCNI in our
`master` CI, so as to keep in sync with the new recommended way for
users to create AKS clusters and install Cilium on them.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
docs: add BYOCNI as preferred option in AKS instructions

Following up on the previous commit: we add documentation related to the
new AKS BYOCNI capabilities, and recommend it as the preferred
installation method for AKS clusters.

Since users might still want to use the integration offered by Azure
IPAM, we elect to keep both the older Azure IPAM-based AKS installation
method and the new BYOCNI-based installation method in the
documentation.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

mattstam commented on Documentation/gettingstarted/k8s-install-default.rst:
```diff
@@ -274,7 +212,18 @@ You can install Cilium on any Kubernetes cluster. Pick one of the options below:
 
-           cilium install --azure-resource-group "${AZURE_RESOURCE_GROUP}"
+           cilium install --azure-resource-group "${AZURE_RESOURCE_GROUP}" \
+             --datapath-mode=tunnel \
```
Is `--datapath-mode=tunnel` equivalent to the helm ` --set tunnel=vxlan`?

nbusseneau replied:
It is! This sets `DatapathMode` to the `DatapathTunnel` value, which itself sets the `tunnel` Helm option to the one defined by `TunnelType`: https://github.com/cilium/cilium-cli/blob/5fcaeb068adbd39229fca62b818175027a6b1df9/install/helm.go#L176-L181

Since `TunnelType` is currently not configurable (not exposed via a CLI `install` flag), it always defaults to `vxlan`: https://github.com/cilium/cilium-cli/blob/5fcaeb068adbd39229fca62b818175027a6b1df9/defaults/defaults.go#L71

mattstam replied:
Thanks! I just want to double check and make sure this is confirmed/tested that these would be the correct configuration options. Previously I was using `--datapath-mode vxlan` to get installation to complete, though some of the Hubble usage might have been unstable.

nbusseneau replied:
Yea I'm 95% sure `--datapath-mode=vxlan` just sets `DatapathMode` to `DatapathTunnel` by default since `vxlan` is not a valid option for `datapath-mode` 😁

nbusseneau replied:
@mattstam FYI we have revamped `DatapathMode` in https://github.com/cilium/cilium-cli/pull/898 and are adding a new BYOCNI-specific mode in https://github.com/cilium/cilium-cli/pull/899, which will be the basis for the main PR at hand here.

nbusseneau committed:
ci: use CLI 0.11.8 for AKS workflow

CLI version 0.11.8 adds support for BYOCNI:
https://github.com/cilium/cilium-cli/releases/tag/v0.11.8

Note that in order to use CLI 0.11.8, the `base-version` flags had to be
removed due to https://github.com/cilium/cilium-cli/pull/890 +
https://github.com/cilium/cilium-cli/pull/891,
and we also had to use manual `kubectl port-forward` instead of
`cilium hubble port-forward` because we seemingly broke that in the
process. This is going to be fixed in a later release of the CLI, but
for now we can get away with it.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
helm: add new `aksbyocni.enabled` integration

Add a new "datapath mode" to the Helm chart for AKS clusters using the
new BYOCNI feature.

This integration is specifically added to avoid confusion with the older
Azure integration, as AKS clusters created in BYOCNI mode cannot use the
Azure IPAM (Azure API not available in BYOCNI mode).

Context:

Microsoft has recently released support for "Bring your own CNI" when
creating AKS clusters, allowing users to create clusters with no CNI
plugin pre-installed.

This is extremely useful for Cilium as previously setting up an AKS
cluster and installing Cilium was very complicated, due to the need to
use multiple nodepools with a complex taint system in order to ensure
applications pods would not get scheduled before Cilium agents were
ready to manage connectivity in the cluster.

The BYOCNI feature has been available in preview as an `az` CLI
extension for some time now, and Microsoft has published a dedicated
documentation page outlining the prequisites and implications (notably
in terms of support policy) of BYOCNI:
https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli

BYOCNI in its current state is a strong enough proposal to warrant a
change to the recommendations we make to our users for newly created AKS
clusters.

Here are the changes necessary to run Cilium on an AKS cluster in BYOCNI
mode:

- Azure IPAM does not work in BYOCNI mode as the Azure API is not
  available. Instead of using the Azure operator, we must use the
  generic operator with Cluster Pool IPAM.
- Direct routing does not work in BYOCNI mode. Instead, we must use
  VXLAN encapsulation.

To this end, we simply create a new Helm integration completely separate
of the previous `azure.enabled`, since the old method and the new BYOCNI
methods are mutually exclusive due to Azure IPAM not being available.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

joestringer commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -191,6 +191,12 @@ updateStrategy:
 
 # Configuration Values for cilium-agent
 
+aksbyocni:
```
After #20066 was merged, now these changes need to be made to `install/kubernetes/cilium/values.yaml.tmpl` instead, then run `make -C install/kubernetes` to update this file.

nbusseneau replied:
Done 👍🏻

nbusseneau committed:
docs: remove "custom taint scenario" for AKS clusters using Azure ipam

Our previously recommended scenario for creating new AKS clusters with
proper taint management is voided due to
https://github.com/Azure/AKS/issues/2934.

We switch to not recommending using taints anymore and instead warn
users about the limits of not using taints and the fact we have no
standard and foolproof alternative to recommend.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

nbusseneau committed:
ci: remove taints from AKS workflows in stable CI

As a follow-up to the previous commit, we remove taints from AKS CI in
stable branches. This basically makes no difference in CI because no
new nodes are added and any unmanaged pods are restarted at `cilium
install` time.

Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>

christarazi commented on install/kubernetes/cilium/README.md:
```diff
@@ -56,10 +56,11 @@ contributors across the globe, there is almost always someone available to help.
 | agent | bool | `true` | Install the cilium agent resources. |
 | agentNotReadyTaintKey | string | `"node.cilium.io/agent-not-ready"` | Configure the key of the taint indicating that Cilium is not ready on the node. When set to a value starting with `ignore-taint.cluster-autoscaler.kubernetes.io/`, the Cluster Autoscaler will ignore the taint on its decisions, allowing the cluster to scale up. |
+| aksbyocni.enabled | bool | `false` | Enable AKS BYOCNI integration. Note that this is incompatible with AKS clusters not created in BYOCNI mode: use Azure integration (`azure.enabled`) instead. |
 | alibabacloud.enabled | bool | `false` | Enable AlibabaCloud ENI integration |
 | annotateK8sNode | bool | `false` | Annotate k8s node upon initialization with Cilium's metadata. |
 | autoDirectNodeRoutes | bool | `false` | Enable installation of PodCIDR routes between worker nodes if worker nodes share a common L2 network segment. |
-| azure.enabled | bool | `false` | Enable Azure integration |
+| azure.enabled | bool | `false` | Enable Azure integration. Note that this is incompatible with AKS clusters created in BYOCNI mode: use AKS BYOCNI integration (`aksbyocni.enabled`) instead. |
```
Can we make `aksbyocni` under `azure` like `azure.byocni`? I don't see a need to have them as separate top-level values.

nbusseneau replied:
I thought about it but decided this was a bad idea:

- Azure IPAM is not usable when using AKS BYOCNI (API not available), both features are mutually exclusive.
- The `azure` scope is historically only used for the Azure integration, and has several knobs that only relate to Azure IPAM, with the main one being `azure.enabled`. Having BYOCNI in there would make it a potential mess, because it makes BYOCNI the only cloud-provider mode that does not adhere to the `<cloud>.enabled` convention, and also you'd either have to:
  - Use `azure.byocni=true` with `azure.enabled=false`, otherwise Azure IPAM and other knobs kick in.
  - Use `azure.byocni=true` with `azure.enabled=true` BUT tweak the chart all around so that every `if` block for `azure.enabled` has `and not azure.byocni=true` guards, e.g. for the Azure IPAM secrets.
- The BYOCNI feature is AKS-specific, and not something Azure-wide, thus bundling it under the `azure` scope does not feel accurate.

nbusseneau replied:
Another option which I thought about was NOT to have a `aksbyocni` group at all, since all it does is set IPAM to Cluster Pool and disable tunneling. So we don't need a new mode at all and could simply have both options passed in correctly to the Helm chart in the docs, and the CLI handling it transparently for the CLI install (as it already does).

But we decided with @aanm that it was nice to advertise BYOCNI as an explicit alternative to Azure IPAM to the Helm values.

christarazi replied:
:ok_hand: Makes sense, thanks for the explanation.

christarazi reviewed (commented):
LGTM, one nit.

kaworu commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -371,6 +371,8 @@ data:
   enable-endpoint-routes: "true"
   enable-local-node-route: "false"
+{{- else if .Values.aksbyocni.enabled }}
+  tunnel: "vxlan"
```
Isn't `geneve` also a valid value here?

aanm replied:
I believe we haven't tested it out. Something for @nbusseneau to look afterwards.

kaworu commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -371,6 +371,8 @@ data:
+{{- else if .Values.aksbyocni.enabled }}
+  tunnel: "vxlan"
 {{- else }}
   tunnel: {{ .Values.tunnel }}
```
I know this line is unchanged by this PR, but since we're here could we `quote` please?

```suggestion
  tunnel: {{ .Values.tunnel | quote }}
```

qmonnet commented on Documentation/gettingstarted/k8s-install-default.rst:
```diff
@@ -56,7 +56,52 @@ to create a Kubernetes cluster locally or using a managed Kubernetes service:
+       for instructions on how to install ``az`` and prepare your account, and
+       the `Bring your own CNI documentation
+       <https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli>`_
+       for more details about BYOCNI prerequisites / implications.
```
```suggestion
       for more details about BYOCNI prerequisites/implications.
```

qmonnet commented on Documentation/gettingstarted/k8s-install-default.rst:
```diff
@@ -56,7 +56,52 @@ to create a Kubernetes cluster locally or using a managed Kubernetes service:
+       .. code-block:: bash
+
+           export NAME="$(whoami)-$RANDOM"
+           export AZURE_RESOURCE_GROUP="${NAME}-group"
```
Are the `export`s needed?

nbusseneau replied:
They are not needed, I just copy/pasted whatever we had in the already existing AKS instructions. I would suggest to do a separate PR to remove all unneeded `export` at once from all GSG if we want to do so, this way we keep consistency in the docs before and after that PR is merged.

qmonnet commented on Documentation/gettingstarted/requirements-aks-byocni.rst:
```diff
@@ -0,0 +1,23 @@
+
+* The AKS cluster must be created with ``--network-plugin none`` (BYOCNI). See
+  the `Bring your own CNI documentation <https://docs.microsoft.com/en-us/azure/aks/use-byo-cni?tabs=azure-cli>`_
+  for more details about BYOCNI prerequisites / implications.
```
```suggestion
  for more details about BYOCNI prerequisites/implications.
```

qmonnet reviewed (approved):
Looks good!

sayboras commented on .github/workflows/conformance-aks.yaml:
```diff
@@ -210,41 +209,11 @@ jobs:
             --location ${{ env.location }} \
-            --network-plugin azure \
-            --node-count 1 \
+            --network-plugin none \
```
I can't find where we set the helm attribute aksbyocni.enabled=true in this workflow

nbusseneau replied:
We don't, we are using the CLI in this workflow, so the CLI is responsible for setting up the Helm values as expected ;)

sayboras replied:
probably related to next comment https://github.com/cilium/cilium/pull/19379#discussion_r900893669, I was comparing CLI command for both BYOCNI and AKS IPAM.

sayboras commented on Documentation/gettingstarted/k8s-install-default.rst:
```diff
@@ -264,9 +309,21 @@ You can install Cilium on any Kubernetes cluster. Pick one of the options below:
+
+       .. code-block:: shell-session
+
+           cilium install --azure-resource-group "${AZURE_RESOURCE_GROUP}"
```
should we have something else to set byocni flag as well ? I think this command is just for AKS ipam mode.

nbusseneau replied:
As noted above: this is handled by the CLI. You can check out https://github.com/cilium/cilium-cli/pull/899 for the full details on the implementation :)

sayboras replied:
oh right, do we still need to pass datapath-mode flag ? :thinking:

sayboras replied:
got it :100: , beautiful work :100:

sayboras commented on Documentation/gettingstarted/k8s-install-helm.rst:
```diff
@@ -81,9 +81,24 @@ Install Cilium
+
+       .. parsed-literal::
+
+          helm install cilium |CHART_RELEASE| \\
+            --namespace kube-system \\
+            --set aksbyocni.enabled=true \\
+            --set nodeinit.enabled=true
```
:100:

This is surprisingly simpler than I thought. Just test it out, working as expected 

```
$ helm install cilium ./cilium \
  --namespace kube-system \
  --set aksbyocni.enabled=true \
  --set nodeinit.enabled=true
NAME: cilium
LAST DEPLOYED: Sat Jun 18 23:16:18 2022
NAMESPACE: kube-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You have successfully installed Cilium with Hubble.

Your release version is 1.11.90.

For any further help, visit https://docs.cilium.io/en/v1.11/gettinghelp

$ cilium status
    /¯¯\
 /¯¯\__/¯¯\    Cilium:         OK
 \__/¯¯\__/    Operator:       OK
 /¯¯\__/¯¯\    Hubble:         disabled
 \__/¯¯\__/    ClusterMesh:    disabled
    \__/

DaemonSet         cilium             Desired: 3, Ready: 3/3, Available: 3/3
Deployment        cilium-operator    Desired: 2, Ready: 2/2, Available: 2/2
Containers:       cilium             Running: 3
                  cilium-operator    Running: 2
Cluster Pods:     4/4 managed by Cilium
Image versions    cilium             quay.io/cilium/cilium-ci:latest: 3
                  cilium-operator    quay.io/cilium/operator-generic-ci:latest: 2

```

sayboras reviewed (commented):
Few comments as per below, the rest looks good to me :100:

kaworu reviewed (commented):
Couple of comments, but Helm changes LGTM.

sayboras reviewed (approved):
LGTM :100:

aanm reviewed (commented):
@qmonnet @nbusseneau @kaworu I'll merge the PR as is. I've address a couple comments from Alex. Let's fix it as a follow up
